### PR TITLE
Boltz.exchange Extension

### PR DESCRIPTION
**Pull request by dni**
boltz.exchange integration, please review and find all the bugs :)

## Todo:
* [x] Make sure to rebase to `main`
* [x] Apply `black` and `isort` to everything
* [x] Write tests
* [x] Add migrations to `conv.py`
* [x] All checks succeed
* [x] Write `README.md`
* [x] Remove comments from code
* [x] startup swap checks

----

i just successfully did a swap and a reverse swap on mainnet :),
to anyone else here you can test it yourself.
my lnbits with boltz (slow vps): https://lnbits.b1tco1n.org
readme: https://github.com/lnbits/lnbits-legend/tree/boltz/lnbits/extensions/boltz
first normal swap: https://mempool.space/address/3M9ULSRJNYs2qPqjQZzyveS4o7at6w1tnm
first reverse swap: https://mempool.space/address/bc1qyh87j7l9s050a9e0fdnwljdzfcafe9s23e06pkgft8qqmkz3a0uqklpkqq

### CODE REVIEW

File: lnbits/extensions/boltz/tasks.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+  3. test: create -> kill -> mine blocks and hit timeout -> start -> should go timeout/failed
+  4. test: create -> kill -> pay to less onchain funds -> mine blocks hit timeout -> start lnbits -> should be refunded
+
+B. reverse swaps
+  1. test: create instant -> kill -> boltz does lockup -> not confirmed -> start lnbits -> should claim/complete
+  2. test: create instant -> kill -> no lockup -> start lnbits -> should start onchain listener -> boltz does lockup -> should claim/complete (difficult to test)
+  3. test: create -> kill -> boltz does lockup -> not confirmed -> start lnbits -> should start tx listener -> after confirmation -> should claim/complete
+  4. test: create -> kill -> boltz does lockup -> confirmed -> start lnbits -> should claim/complete
+  5. test: create -> kill -> boltz does lockup -> hit timeout -> boltz refunds -> start -> should timeout
+"""
```
Comment (callebtc) [inline lnbits/extensions/boltz/tasks.py:39]: Why are these here? Shouldn't it be in `/tests/` rather?

File: lnbits/extensions/boltz/tasks.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+                if payment_status.paid:
+                    logger.debug(
+                        f"Boltz - swap: {swap.boltz_id} got paid while offline."
+                    )
+                    await update_swap_status(swap.id, "complete")
+                else:
+                    if swap_status.hit_timeout:
+                        if not swap_status.has_lockup:
+                            logger.debug(
```
Comment (callebtc) [inline lnbits/extensions/boltz/tasks.py:76]:
```suggestion
                            logger.warning(
```

File: lnbits/extensions/boltz/tasks.py (hunk 1 above)
Comment (dni) [reply lnbits/extensions/boltz/tasks.py:39]: currently i have no good way to test the startup function, so i put i there for me to remember for manual testing

File: lnbits/extensions/boltz/boltz.py [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+"""
+explanation taken from electrum
+send on Lightning, receive on-chain
+- User generates preimage, RHASH. Sends RHASH to server.
+- Server creates an LN invoice for RHASH.
+- User pays LN invoice - except server needs to hold the HTLC as preimage is unknown.
+- Server creates on-chain output locked to RHASH.
+- User spends on-chain output, revealing preimage.
+- Server fulfills HTLC using preimage.
+Note: expected_onchain_amount_sat is BEFORE deducting the on-chain claim tx fee.
```
Comment (callebtc) [inline lnbits/extensions/boltz/boltz.py:108]: I don't understand this note.

File: lnbits/extensions/boltz/boltz.py [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+    if not check_boltz_limits(data.amount):
+        msg = f"Boltz - reverse swap not in boltz limits"
+        logger.warning(msg)
+        raise Exception(msg)
+
+    swap_id = urlsafe_short_hash()
+
+    if not await check_balance(data):
+        logger.error(
+            f"Boltz - reverse swap, insufficient balance. create_reverse_swap should never be called before balance checking."
```
Comment (callebtc) [inline lnbits/extensions/boltz/boltz.py:125]:
```suggestion
            f"Boltz - reverse swap, insufficient balance."
```

This error is user-facing, it doesn't really make sense to tell them "create_reverse_swap should never be called before balance checking." – why not put the balance check into create_reverse_swap?

Comment (dni) [reply lnbits/extensions/boltz/boltz.py:125]: my idea behind this was, although it maybe not a good idea. the user has only the api exposed to him, where i do the balance check. but if so another ext or anything else uses it (like tests) i double check it and give this error.

i think i will change that, i was an afterthought when i was writing tests, if i use it in both cases i might aswell only check there.

File: lnbits/extensions/boltz/views_api.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+        405: {"description": "not allowed method, insufficient balance"},
+        500: {"description": "boltz error"},
+    },
+)
+async def api_submarineswap_create(
+    data: CreateSubmarineSwap,
+    wallet: WalletTypeInfo = Depends(require_admin_key),  # type: ignore
+):
+    try:
+        data = await create_swap(data)
```
Comment (callebtc) [inline lnbits/extensions/boltz/views_api.py:152]: You're overwriting `data` with a variable of a different type. Mypy didn't complain?

File: lnbits/extensions/boltz/boltz.py (hunk 3 above)
Comment (dni) [reply lnbits/extensions/boltz/boltz.py:108]: that note means that you need to calculate the fees when doing a reverse swap and make it so expected_onchain_amount_sat - fees = amount yo need to claim.

File: lnbits/extensions/boltz/views_api.py (hunk 5 above)
Comment (dni) [reply lnbits/extensions/boltz/views_api.py:152]: no because, https://github.com/lnbits/lnbits-legend/pull/873, is not yet in :)

File: lnbits/extensions/boltz/views_api.py [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+async def api_submarineswap_status(
+    swap_id: str, wallet: WalletTypeInfo = Depends(require_admin_key)  # type: ignore
+):
+    swap = await get_submarine_swap(swap_id)
+    if swap == None:
+        swap = await get_reverse_submarine_swap(swap_id)
+        if swap == None:
+            raise HTTPException(
+                status_code=HTTPStatus.NOT_FOUND, detail="swap does not exist."
+            )
```
Comment (callebtc) [inline lnbits/extensions/boltz/views_api.py:245]:
```suggestion
    swap = await get_submarine_swap(swap_id) or await get_reverse_submarine_swap(swap_id)
    if swap == None:
        raise HTTPException(
                status_code=HTTPStatus.NOT_FOUND, detail="swap does not exist."
            )
```

Comment (dni) [reply lnbits/extensions/boltz/views_api.py:245]: beautiful

File: lnbits/extensions/boltz/boltz.py (hunk 4 above)
Comment (dni) [reply lnbits/extensions/boltz/boltz.py]: 31546088

File: lnbits/extensions/boltz/views_api.py (hunk 5 above)
Comment (dni) [reply lnbits/extensions/boltz/views_api.py]: changes it: 3f717ea3

File: lnbits/extensions/boltz/boltz.py [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+    if swap_status.block_height >= swap.timeout_block_height:
+        swap_status.hit_timeout = True
+
+    mempool_tx = get_mempool_tx(swap_status.address)
+    swap_status.lockup = mempool_tx
+    if mempool_tx == None:
+        swap_status.has_lockup = False
+        swap_status.confirmed = False
+        swap_status.mempool = "transaction.unknown"
+        swap_status.message = "lockup tx not in mempool"
```
Comment (michael1011) [inline lnbits/extensions/boltz/boltz.py:392]: Keep in mind that people could still send lockup transactions for normal swaps at some random point in time after it was created (users sometimes do weird stuff)

File: lnbits/extensions/boltz/boltz.py [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+
+    if swap_status.block_height >= swap.timeout_block_height:
+        swap_status.hit_timeout = True
+
+    mempool_tx = get_mempool_tx(swap_status.address)
+    swap_status.lockup = mempool_tx
+    if mempool_tx == None:
+        swap_status.has_lockup = False
+        swap_status.confirmed = False
+        swap_status.mempool = "transaction.unknown"
```
Comment (michael1011) [inline lnbits/extensions/boltz/boltz.py:391]: Btw: an enum for the status would be nice

File: lnbits/extensions/boltz/crud.py [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+
+async def get_submarine_swap(swap_id) -> SubmarineSwap:
+    row = await db.fetchone(
+        "SELECT * FROM boltz.submarineswap WHERE id = ?", (swap_id,)
+    )
+    return SubmarineSwap(**row) if row else None
+
+
+async def create_submarine_swap(swap: SubmarineSwap) -> Optional[SubmarineSwap]:
+
```
Comment (michael1011) [inline lnbits/extensions/boltz/crud.py:62]: Empty line

File: lnbits/extensions/boltz/crud.py [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+            swap.address,
+            swap.bip21,
+            swap.redeem_script,
+            swap.amount,
+        ),
+    )
+    return await get_submarine_swap(swap.id)
+
+
+async def delete_submarine_swap(swap_id):
```
Comment (michael1011) [inline lnbits/extensions/boltz/crud.py:101]: Unused method. And I don't think you should ever delete any kind of swap information from the database

File: lnbits/extensions/boltz/crud.py [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+            swap.timeout_block_height,
+            swap.redeem_script,
+            swap.amount,
+        ),
+    )
+    return await get_reverse_submarine_swap(swap.id)
+
+
+async def update_swap_status(swap_id: str, status: str):
+
```
Comment (michael1011) [inline lnbits/extensions/boltz/crud.py:197]: Empty line

File: lnbits/extensions/boltz/mempool.py [hunk 12]
```diff
@@ -0,0 +1,97 @@
+import asyncio
```
Comment (michael1011) [inline lnbits/extensions/boltz/mempool.py:1]: This file is in the Boltz extension. Wouldn't it be useful other extensions too if we make a little more generic?

File: lnbits/extensions/boltz/tasks.py [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+  1. test: create instant -> kill -> boltz does lockup -> not confirmed -> start lnbits -> should claim/complete
+  2. test: create instant -> kill -> no lockup -> start lnbits -> should start onchain listener -> boltz does lockup -> should claim/complete (difficult to test)
+  3. test: create -> kill -> boltz does lockup -> not confirmed -> start lnbits -> should start tx listener -> after confirmation -> should claim/complete
+  4. test: create -> kill -> boltz does lockup -> confirmed -> start lnbits -> should claim/complete
+  5. test: create -> kill -> boltz does lockup -> hit timeout -> boltz refunds -> start -> should timeout
+"""
+
+
+async def check_for_pending_swaps():
+
```
Comment (michael1011) [inline lnbits/extensions/boltz/tasks.py:43]: Empty line

File: lnbits/extensions/boltz/tasks.py [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+    invoice_queue = asyncio.Queue()
+    register_invoice_listener(invoice_queue)
+
+    while True:
+        payment = await invoice_queue.get()
+        await on_invoice_paid(payment)
+
+
+async def on_invoice_paid(payment: Payment) -> None:
+    if "boltz" != payment.extra.get("tag"):
```
Comment (michael1011) [inline lnbits/extensions/boltz/tasks.py:139]: A constant for `boltz` would be nice

File: lnbits/extensions/boltz/tasks.py [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+        # not a boltz invoice
+        return
+
+    await payment.set_pending(False)
+    swap_id = payment.extra.get("swap_id")
+    swap = await get_submarine_swap(swap_id)
+
+    if not swap:
+        logger.error(f"swap_id: {swap_id} not found. updating status failed.")
+        return
```
Comment (michael1011) [inline lnbits/extensions/boltz/tasks.py:149]: What `update status failed` are you doing here?

File: lnbits/extensions/boltz/boltz.py [hunk 16]
```diff
@@ -0,0 +1,430 @@
...
+)
+from .utils import check_balance, get_timestamp, req_wrap
+
+net = NETWORKS[BOLTZ_NETWORK]
+logger.debug(f"BOLTZ_URL: {BOLTZ_URL}")
+logger.debug(f"Bitcoin Network: {net['name']}")
+
+
+async def create_swap(data: CreateSubmarineSwap) -> SubmarineSwap:
+
```
Comment (michael1011) [inline lnbits/extensions/boltz/boltz.py:42]: Empty line

**Review by michael1011 — CHANGES_REQUESTED** [review]
Looks good in general. Just needs some final touches

File: lnbits/extensions/boltz/mempool.py (hunk 12 above)
Comment (dni) [reply lnbits/extensions/boltz/mempool.py:1]: i am planning on doing a mempool extension, which will take care of things like that.

File: lnbits/extensions/boltz/tasks.py (hunk 14 above)
Comment (dni) [reply lnbits/extensions/boltz/tasks.py:139]: currently every extension does that. so for consistency i will keep that.

File: lnbits/extensions/boltz/mempool.py (hunk 12 above)
Comment (dni) [reply lnbits/extensions/boltz/mempool.py:1]: there is this proposal of mine i did in april: https://github.com/lnbits/lnbits-legend/issues/584

**Review by callebtc — APPROVED** [review]
LGTM

**Review by callebtc — APPROVED** [review]
